### ISSUE TITLE
Fix false positives for variables in function-calc-no-invalid

### ISSUE THREAD

**Issue by vilchik-elena** [stylelint/stylelint#4114]
There are false positives for less and scss variables containing `-`

> Which rule, if any, is the bug related to?

`function-calc-no-invalid`

> What CSS is needed to reproduce the bug?

for less

```css
.foo {
	height: calc(100% - (@a-height + @b-margin-top));
}
```
for scss
```css
.foo {
	height: calc(100% - ($a-height + $b-margin-top));
}
```

> What stylelint configuration is needed to reproduce the bug?
```json
{
  "rules": {
    "function-calc-no-invalid": true
  }
}
```

> Which version of stylelint are you using?

`10.1.0`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

CLI with `stylelint file.less --config myconfig.json`

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

Yes, it's related to less and scss variables

> What did you expect to happen?

No warnings to be flagged

> What actually happened (e.g. what warnings or errors did you get)?

The following warnings were flagged (for less case), a pair of issues per dash in a variable name

```shell
 2:25  ✖  Expected space before "-" operator   function-calc-no-invalid
 2:26  ✖  Expected space after "-" operator    function-calc-no-invalid
 2:37  ✖  Expected space before "-" operator   function-calc-no-invalid
 2:38  ✖  Expected space after "-" operator    function-calc-no-invalid
 2:44  ✖  Expected space before "-" operator   function-calc-no-invalid
 2:45  ✖  Expected space after "-" operator    function-calc-no-invalid
```

There was a bug fix done (#3908) to fix variables inside interpolation for both less and scss, but for simple variable there are still FPs. Note that for scss as soon as I wrap variables inside interpolation, issues disappear :
```
.foo {
	height: calc(100% - (#{$a-height} + #{$b-margin-top}));
}
```

**Comment by hudochenkov** [comment]
Thanks for the report and for using the template.

Looks like it's the same problem as https://github.com/stylelint/stylelint/issues/4102. @ota-meshi could you take a look, please?

**Comment by ota-meshi** [comment]
Hello!

I checked about `SCSS`.
In the case of `SCSS`, I think that there is no problem in the current behavior.

When I tried [here](https://www.sassmeister.com/), the following `SCSS` syntax seems to be invalid:

```scss
$a-height: 10px;
$b-margin-top: 20px;

.foo {
  height: calc(100% - ($a-height + $b-margin-top));
}
```

The following SCSS syntax is valid:

```scss
$a-height: 10px;
$b-margin-top: 20px;

.foo {
  height: calc(100% - (#{$a-height} + #{$b-margin-top}));
}
```

In this syntax, there is no problem with the `function-calc-no-invalid` rule.

I checked about `less`.
In the case of `less`, as some of you said, there is a problem with the current behavior.

When I tried [here](http://lesscss.org/less-preview/), the following `less` syntax is valid:

```less
@a-height: 10px;
@b-margin-top: 20px;

.foo {
	height: calc(100% - (@a-height + @b-margin-top));
}
```

In this syntax, there is false positive in the `function-calc-no-invalid` rule.

This is a bug but I'm not familiar with `less`.
Please let me know if you know the character codes available for `less` variables.

**Comment by xavierfoucrier** [comment]
Hi everyone,

This bug is still present with the latest **stylelint@11.0.0** release: you can fix it using `@{variable-name}` syntax.

> Note that the `function-calc-no-invalid` error wasn't thrown with **stylelint@10.1.0**.

Hope this help! :wink:

**Comment by xavierfoucrier** [comment]
Hi everyone,

This bug is still present with the latest **stylelint@11.0.0** release...

> Note that the `function-calc-no-invalid` error wasn't thrown with **stylelint@10.1.0**.

Hope this help! :wink:

**Comment by LolliDepp** [comment]
Still an issue as of 12.0.1

**Comment by xavierfoucrier** [comment]
I confirm that it's still an issue on **stylelint@12.0.1**.

**Comment by LolliDepp** [comment]
Opened PR #4534 to highlight underlying issues in the test cases

**Comment by jeddy3** [comment]
As the issue has been labelled "help wanted", please consider [contributing a fix](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#fixing-a-bug-in-an-existing-rule) if anyone has time.

**Comment by jeddy3** [comment]
@ota-meshi  said:

> This is a bug but I'm not familiar with less.
> Please let me know if you know the character codes available for less variables.

Built-in rules ignore non-standard syntax. This rule should only check the validity of `calc` functions containing standard CSS syntax.

i.e. it should return early if there is any `$`, `@`, `{` or `}` characters within the `calc` function. 

We might be able to use the [`isStandardSyntax*` utils](https://github.com/stylelint/stylelint/tree/master/lib/utils) or it could be an occasion where we need to use bespoke heuristics.

**Comment by jeddy3** [comment]
Looking [at the specs](https://www.w3.org/TR/css-values-4/#calc-notation) we'll probably want to create a `isStandardSyntaxMathFunction` util that returns `false` if it encounters any `$`, `@`, `{` or `}` characters.

See the [`isStandardSyntaxMediaFeature` util](https://github.com/stylelint/stylelint/blob/master/lib/utils/isStandardSyntaxMediaFeature.js) for prior art.

**Comment by jeddy3** [comment]
I've labeled the issue as ready to implement using the `isStandardSyntaxMathFunction` approach. Please consider [contributing](https://github.com/stylelint/stylelint/blob/master/CONTRIBUTING.md) if you have time.

There are [steps on how to fix a bug in a rule](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by luckylooke** [comment]
Hello 👋

Hypens seems to be the problem here
```
  ✖ Expected space after "-" operator (function-calc-no-invalid)
  729 |           border-left: none;
  730 |           top: calc(#{$ws-thick-line} * -1);
> 731 |           height: calc(100% + calc(#{$ws-thick-line} * 2));
      |                                                ^
  732 |         }
  733 |       }
  734 |     }
```
also multiply by -1 is issue
```
  ✖ Expected a valid expression (function-calc-no-invalid)
  728 |           // border-radius: $ws-border-radius; // commented out reason, see above
  729 |           border-left: none;
> 730 |           top: calc(-1 * #{$xxxxxx});
      |                     ^
  731 |           height: calc(100% + calc(#{$xxxxxx} * 2));
  732 |         }
  733 |       }
```

**info**
Vue CLI 

- environment
- running command `vue-cli-service lint:style --syntax scss`
- stylelint version:
```
npm list stylelint
cart@1.0.0 /mnt/AA36F40936F3D473/projects/cart-client
├─┬ @ascendancyy/vue-cli-plugin-stylelint@1.1.2
│ └── stylelint@9.10.1
└── stylelint@12.0.1
```
